urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/laazotea/indi/driver/INDIFilterWheelDriver.java

```diff
@@ -0,0 +1,222 @@
+   * Indicates how many filters does the filter wheel manage.
+   *
+   * @return The number of filters that the filter wheel manages.
+   */
+  public abstract int getNumberOfFilters();
+
+  /**
+   * Constructs a INDIFilterWheelDriver with a particular
+   * <code>inputStream<code> from which to read the incoming messages (from clients) and a
+   * <code>outputStream</code> to write the messages to the clients.
+   *
+   * @param inputStream The stream from which to read messages
+   * @param outputStream The stream to which to write the messages
+   */
+  public INDIFilterWheelDriver(InputStream inputStream, OutputStream outputStream) {
+    super(inputStream, outputStream);
+  }
+
+  /**
+   * Initializes the standard properties. MUST BE CALLED BY SUBDRIVERS.
+   */
+  protected void initializeStandardProperties() {
+    try {
+      filterNamesP = (INDITextProperty)INDIProperty.loadFromFile(this, "filter_names");
+    } catch (INDIException ex) {
+      System.out.println(ex.getMessage());
+    }
+
+    if (filterNamesP == null) {
+      filterNamesP = INDITextProperty.createSaveableTextProperty(this, "filter_names", "Filter Names", "Configuration", Constants.PropertyStates.OK, Constants.PropertyPermissions.RW, 0);
+
+      for (int i = 0 ; i < getNumberOfFilters() ; i++) {
+        INDITextElement te = new INDITextElement(filterNamesP, "filter_name_" + (i + 1), "Filter " + (i + 1), "Filter " + (i + 1));
+      }
+    }
+
+    filterSlotP = new INDINumberProperty(this, "FILTER_SLOT", "Filter Slot", "Control", Constants.PropertyStates.IDLE, Constants.PropertyPermissions.RW, 0);
+    filterSlotValueE = new INDINumberElement(filterSlotP, "FILTER_SLOT_VALUE", "Filter Slot Value", 1, 1, getNumberOfFilters(), 1, "%1.0f");
+
+    filterNameP = new INDITextProperty(this, "FILTER_NAME", "Filter Name", "Control", Constants.PropertyStates.IDLE, Constants.PropertyPermissions.RO, 0);
+    String firstFilterName = filterNamesP.getElement("filter_name_1").getValue();
+    filterNameValueE = new INDITextElement(filterNameP, "FILTER_NAME_VALUE", "Filter Name Value", firstFilterName);
+
+    addProperty(filterNamesP);
+  }
+
+  @Override
+  public void processNewTextValue(INDITextProperty property, Date timestamp, INDITextElementAndValue[] elementsAndValues) {
+    if (property == filterNamesP) {
+      for (int i = 0 ; i < elementsAndValues.length ; i++) {
+        INDITextElement el = elementsAndValues[i].getElement();
+        String val = elementsAndValues[i].getValue();
+        el.setValue(val);
+      }
+
+      filterNamesP.setState(Constants.PropertyStates.OK);
+
+      try {
+        updateProperty(filterNamesP);
+      } catch (INDIException e) {
+        e.printStackTrace();
+      }
+    }
+  }
+
+  @Override
+  public void processNewNumberValue(INDINumberProperty property, Date timestamp, INDINumberElementAndValue[] elementsAndValues) {
+    if (property == filterSlotP) {
+      int newFilterNumber = elementsAndValues[0].getValue().intValue();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;elementsAndValues&#39; (index: 2 | type: INDINumberElementAndValue[). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;property&#39;. The same parameter type is validated here for example: urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/farom/astroiddriver/INDIAstroidDriver.java:387. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).